### PR TITLE
Add Lightning paper benchmarks

### PR DESCRIPTION
**Pull request by AmintorDusko**
This PR adds three specially selected Lightning paper workflows, with support for `lightning.qubit` and `default.qubit`.

### CODE REVIEW

**Comment by AmintorDusko** [comment]
[[sc-60587]](https://app.shortcut.com/xanaduai/story/60587/integrate-at-least-three-existing-performance-paper-benchmarks-into-the-benchmark-suite)

File: benchmarks/conftest.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    """Return the pennylane device.
+
+    The device is ``lightning.qubit`` by default. Allowed values are:
+    "default.qubit", and "lightning.qubit". An
+    underscore can also be used instead of a dot. If the environment
+    variable ``PL_DEVICE`` is defined, its value is used. Underscores
+    are replaced by dots upon exiting.
+    """
+    device = None
+    if "DEVICE" in os.environ:
```
Comment (jay-selby) [inline benchmarks/conftest.py:37]: `DEVICE` or `PL_DEVICE`?  The above comment mentions `PL_DEVICE`.

**Review by jay-selby — CHANGES_REQUESTED** [review]

File: benchmarks/conftest.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+Pytest configuration file for PennyLane-Benchmarks suite.
+"""
+import os
+
+import pennylane as qml
+
+# Looking for the device for benchmarking.
+default_device = "lightning.qubit"
+supported_devices = {"default.qubit", "lightning.qubit"}
+supported_devices.update({sb.replace(".", "_") for sb in supported_devices})
```
Comment (maliasadi) [inline benchmarks/conftest.py:24]: Why do you need this device string update?

```suggestion
supported_devices = {"default.qubit", "lightning.qubit"}
```

File: benchmarks/conftest.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    The device is ``lightning.qubit`` by default. Allowed values are:
+    "default.qubit", and "lightning.qubit". An
+    underscore can also be used instead of a dot. If the environment
+    variable ``PL_DEVICE`` is defined, its value is used. Underscores
+    are replaced by dots upon exiting.
+    """
+    device = None
+    if "DEVICE" in os.environ:
+        device = os.environ.get("DEVICE", default_device)
+        device = device.replace("_", ".")
```
Comment (maliasadi) [inline benchmarks/conftest.py:39]: The same here,

```suggestion
```

File: benchmarks/QAOA/common_utils.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
+from typing import List, Optional, Tuple
+import pennylane as qml
+import networkx as nx
+from pennylane import numpy as np
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply benchmarks/QAOA/common_utils.py:4]: I didn't add isort here. But I can do it.

File: benchmarks/conftest.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        raise ValueError(f"Invalid backend {device}.")
+    return device
+
+
+device_name = get_device()
+
+
+def create_device(device_name, wires, shots=None):
+    if device_name == "lightning.qubit":
+        return qml.device(device_name, wires=wires, batch_obs=True, shots=shots)
```
Comment (maliasadi) [inline benchmarks/conftest.py:52]: do we also need to use `grad_on_execution=False` in `qml.qnode` when grad is not required?

File: pennylane_benchmarks/benchmarks_suite_3.11.png
Comment (maliasadi) [inline pennylane_benchmarks/benchmarks_suite_3.11.png:1]: What is the base of this regression/improvements? What do we compare here (PL versions, commits, or devices)?

File: benchmarks/conftest.py (hunk 1 above)
Comment (AmintorDusko) [reply benchmarks/conftest.py:37]: DEVICE. Thank you for that.

File: pennylane_benchmarks/all_benchmarks_xubm.json
Comment (maliasadi) [inline pennylane_benchmarks/all_benchmarks_xubm.json:1]: Why do we need to store the json file of results? Isn't the csv file enough?

File: .github/workflows/run_pennylane_benchmarks.yml
Comment (maliasadi) [inline .github/workflows/run_pennylane_benchmarks.yml:1]: How's this workflow triggered?

**Review by maliasadi — COMMENTED** [review]
Thanks @AmintorDusko! I've had a few general questions, but looks good to merge otherwise :)

File: benchmarks/VQE/test_VQE.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    ["He2", "6-31G"],  # 8 / 181
+    ["H2", "6-31G"],  # 8 / 185
+    # # ["H4", "STO-3G"],  # 8 / 185
+    # ["LiH", "STO-3G"],  # 12 / 631
+    # # ["OH-", "STO-3G"],  # 12 / 631
+    # ["H3+", "6-31G"],  # 12 / 1403
+    # ["BeH2", "STO-3G"],  # 14 / 666
+    # ["H2O", "STO-3G"],  # 14 / 1086
+    # "H2",  # 20 / 2951
+    # ["C2H2", "STO-3G"],  # 24 / 6500
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply benchmarks/VQE/test_VQE.py:20]: Expand the benchmark as required. Increase the workload.

File: benchmarks/conftest.py (hunk 2 above)
Comment (AmintorDusko) [reply benchmarks/conftest.py:24]: So one can call with both 'lightning_qubit' and 'lightning.qubit', this is particularly useful when running workflows that also build Lightning.

File: benchmarks/conftest.py (hunk 3 above)
Comment (AmintorDusko) [reply benchmarks/conftest.py:39]: Same answer as before.

File: benchmarks/QAOA/common_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jay-selby) [reply benchmarks/QAOA/common_utils.py:4]: Not an issue for this round.  Sometime later is fine

File: pennylane_benchmarks/benchmarks_suite_3.11.png
Comment (AmintorDusko) [reply pennylane_benchmarks/benchmarks_suite_3.11.png:1]: The benchmark structure we have in place now can compare commits and/or versions, with no problem.
In the very specific [configuration](https://github.com/PennyLaneAI/pennylane-benchmarks/blob/main/.github/workflows/run_pennylane_benchmarks.yml#L2-L6) we have in place now we compare v0.35.1 and master. Now that we released I think we can update to v0.36.0.
The runners for now are now extremely reliable, though.

File: benchmarks/conftest.py (hunk 5 above)
Comment (AmintorDusko) [reply benchmarks/conftest.py:52]: We could add, but I don't think it is needed. We just need to be consistent with the runs we are comparing.

File: .github/workflows/run_pennylane_benchmarks.yml
Comment (AmintorDusko) [reply .github/workflows/run_pennylane_benchmarks.yml:1]: The workflow can be triggered by workflow_call, schedule and pull_request.
You can find more information [here](https://github.com/PennyLaneAI/pennylane-benchmarks/blob/main/.github/workflows/run_pennylane_benchmarks.yml#L2-L6).

**Review by jay-selby — APPROVED** [review]
LGTM

**Review by maliasadi — APPROVED** [review]
:raised_hands:

**Pull request merged**